File: sdk/python/lightlink/service.py
"""LightLink 服务端模块"""
import asyncio
import json
import uuid
import ssl
from typing import Dict, Callable, Any, Optional
from datetime import datetime
import logging

from nats.aio.client import Client as NATSClient
from lightlink.types import RPCRequest, RPCResponse
from lightlink.metadata import ServiceMetadata, MethodMetadata
from lightlink.client import discover_server_certs, create_ssl_context_from_discovery

logger = logging.getLogger(__name__)

RPCHandler = Callable[[Dict[str, Any]], Any]


class Service:
    """LightLink 服务端"""

    DEFAULT_HEARTBEAT_INTERVAL = 30

    def __init__(
        self,
        name: str,
        nats_url: str = "nats://localhost:4222",
        tls_config: Optional[dict] = None,
        auto_tls: bool = False
    ):
        """
        Initialize service.

        Args:
            name: Service name
            nats_url: NATS server URL
            tls_config: TLS configuration dictionary
            auto_tls: Whether to auto-discover server TLS certificates (mutually exclusive with tls_config)
        """
        if auto_tls and tls_config:
            raise ValueError("Cannot specify both auto_tls and tls_config")

        self.name = name
        self.nats_url = nats_url
        self.tls_config = tls_config
        self.auto_tls = auto_tls
        self.nc: Optional[NATSClient] = None
        self._rpc_handlers: Dict[str, RPCHandler] = {}
        self._method_metadata: Dict[str, MethodMetadata] = {}
        self._heartbeat_task: Optional[asyncio.Task] = None
        self._heartbeat_stop = asyncio.Event()
        self._running = False

    async def register_rpc(self, method: str, handler: RPCHandler) -> None:
        """注册 RPC 方法"""
        self._rpc_handlers[method] = handler
        logger.info(f"Registered RPC method: {method}")

    async def register_method_with_metadata(
        self,
        method: str,
        handler: RPCHandler,
        metadata: MethodMetadata
    ) -> None:
        """注册带元数据的 RPC 方法"""
        self._method_metadata[method] = metadata
        await self.register_rpc(method, handler)

    async def has_rpc(self, method: str) -> bool:
        """检查方法是否已注册"""
        return method in self._rpc_handlers

    async def start(self) -> None:
        """启动服务"""
        if self._running:
            raise RuntimeError("Service already running")

        self.nc = NATSClient()

        # Handle TLS configuration
        if self.auto_tls:
            # Auto-discover server certificates (skip verify for self-signed certs)
            discovery_result = discover_server_certs()
            ssl_ctx = create_ssl_context_from_discovery(discovery_result, verify=False)
            await self.nc.connect(
                self.nats_url,
                tls=ssl_ctx,
                connect_timeout=10,
                reconnect_time_wait=2,
                max_reconnect_attempts=5
            )
        elif self.tls_config:
            # Use provided TLS configuration
            await self.nc.connect(
                self.nats_url,
                tls=self.tls_config,
                connect_timeout=10,
                reconnect_time_wait=2,
                max_reconnect_attempts=5
            )
        else:
            # No TLS
            await self.nc.connect(
                self.nats_url,
                connect_timeout=10,
                reconnect_time_wait=2,
                max_reconnect_attempts=5
            )

        subject = f"$SRV.{self.name}.>"
        await self.nc.subscribe(subject, cb=self._handle_rpc)

        await self._start_heartbeat()
        self._running = True
        logger.info(f"Service '{self.name}' started")

    async def _handle_rpc(self, msg) -> None:
        """处理 RPC 请求"""
        try:
            request_data = json.loads(msg.data.decode())
            request = RPCRequest(**request_data)

            handler = self._rpc_handlers.get(request.method)
            if handler is None:
                await self._send_error(msg, request.id, f"Method not found: {request.method}")
                return

            result = await handler(request.args)
            await self._send_success(msg, request.id, result)

        except Exception as e:
            logger.error(f"Error handling RPC: {e}")
            await self._send_error(msg, "", str(e))

    async def _send_success(self, msg, request_id: str, result: Dict[str, Any]) -> None:
        """发送成功响应"""
        response = RPCResponse(id=request_id, success=True, result=result)
        await msg.respond(json.dumps(response.__dict__).encode())

    async def _send_error(self, msg, request_id: str, error: str) -> None:
        """发送错误响应"""
        response = RPCResponse(id=request_id, success=False, error=error)
        await msg.respond(json.dumps(response.__dict__).encode())

    async def _start_heartbeat(self) -> None:
        """启动心跳"""
        async def heartbeat_loop():
            while not self._heartbeat_stop.is_set():
                await self._send_heartbeat()
                try:
                    await asyncio.wait_for(
                        self._heartbeat_stop.wait(),
                        timeout=self.DEFAULT_HEARTBEAT_INTERVAL
                    )
                except asyncio.TimeoutError:
                    continue

        self._heartbeat_task = asyncio.create_task(heartbeat_loop())

    async def _send_heartbeat(self) -> None:
        """发送心跳"""
        heartbeat = {
            "service": self.name,
            "version": "1.0.0",
            "timestamp": int(datetime.utcnow().timestamp())
        }
        subject = f"$LL.heartbeat.{self.name}"
        await self.nc.publish(subject, json.dumps(heartbeat).encode())

    def build_current_metadata(
        self,
        version: str,
        description: str,
        author: str,
        tags: list[str]
    ) -> ServiceMetadata:
        """构建当前服务的元数据"""
        methods = list(self._method_metadata.values())

        return ServiceMetadata(
            name=self.name,
            version=version,
            description=description,
            author=author,
            tags=tags,
            methods=methods,
            registered_at=datetime.utcnow(),
            last_seen=datetime.utcnow()
        )

    async def register_metadata(self, metadata: ServiceMetadata) -> None:
        """注册服务元数据到 $LL.register.{service}"""
        msg = {
            "service": self.name,
            "version": metadata.version,
            "metadata": metadata.to_dict(),
            "timestamp": int(datetime.utcnow().timestamp())
        }

        subject = f"$LL.register.{self.name}"
        await self.nc.publish(subject, json.dumps(msg).encode())

    async def stop(self) -> None:
        """停止服务"""
        if not self._running:
            return

        self._heartbeat_stop.set()
        if self._heartbeat_task:
            await self._heartbeat_task

        if self.nc:
            await self.nc.close()

        self._running = False
        logger.info(f"Service '{self.name}' stopped")

File: sdk/csharp/LightLink/Service.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using NATS.Client;
using System.Text.Json;
using LightLink.Types;
using LightLink.Metadata;
using System.Security.Cryptography.X509Certificates;
using NATS.Client.Internals;
using System.Net;
using System.Net.NetworkInformation;
using System.Net.Sockets;
using System.Linq;

namespace LightLink
{
    public delegate Task<Dictionary<string, object>> RPCHandler(Dictionary<string, object> args);

    public class Service : IDisposable
    {
        private readonly string _name;
        private readonly string _natsURL;
        private readonly Options? _natsOptions;
        private readonly TLSConfig? _tlsConfig;
        private IConnection? _nc;
        private readonly Dictionary<string, RPCHandler> _rpcHandlers;
        private readonly ReaderWriterLockSlim _rpcLock;
        private readonly Dictionary<string, MethodMetadata> _methodMetadata;
        private readonly ReaderWriterLockSlim _metaLock;
        private Timer? _heartbeatTimer;
        private bool _running;

        private const int HeartbeatIntervalMs = 30000;

        public Service(string name, string natsURL) : this(name, natsURL, null, null)
        {
        }

        public Service(string name, string natsURL, Options natsOptions) : this(name, natsURL, natsOptions, null)
        {
        }

        public Service(string name, string natsURL, TLSConfig tlsConfig) : this(name, natsURL, null, tlsConfig)
        {
        }

        public Service(string name, string natsURL, Options? natsOptions, TLSConfig? tlsConfig)
        {
            _name = name;
            _natsURL = natsURL;
            _natsOptions = natsOptions;
            _tlsConfig = tlsConfig;
            _rpcHandlers = new Dictionary<string, RPCHandler>();
            _methodMetadata = new Dictionary<string, MethodMetadata>();
            _rpcLock = new ReaderWriterLockSlim();
            _metaLock = new ReaderWriterLockSlim();
        }

        public string Name => _name;

        public void RegisterRPC(string method, RPCHandler handler)
        {
            _rpcLock.EnterWriteLock();
            try { _rpcHandlers[method] = handler; }
            finally { _rpcLock.ExitWriteLock(); }
        }

        public void RegisterMethodWithMetadata(string method, RPCHandler handler, MethodMetadata metadata)
        {
            _metaLock.EnterWriteLock();
            try { _methodMetadata[method] = metadata; }
            finally { _metaLock.ExitWriteLock(); }
            RegisterRPC(method, handler);
        }

        public bool HasRPC(string method)
        {
            _rpcLock.EnterReadLock();
            try { return _rpcHandlers.ContainsKey(method); }
            finally { _rpcLock.ExitReadLock(); }
        }

        public void Start()
        {
            if (_running) throw new InvalidOperationException("Service already running");

            var opts = _natsOptions ?? ConnectionFactory.GetDefaultOptions();
            if (opts.Url == null) opts.Url = _natsURL;
            opts.Name = $"LightLink Service: {_name}";

            // Configure TLS if tlsConfig is provided
            if (_tlsConfig != null)
            {
                opts.Secure = true;
                try
                {
                    X509Certificate2 cert;

                    // Prefer PFX format (NATS.Client requires cert + key in same file)
                    if (!string.IsNullOrEmpty(_tlsConfig.PfxFile))
                    {
                        cert = new X509Certificate2(_tlsConfig.PfxFile, _tlsConfig.PfxPassword);
                    }
                    else
                    {
                        cert = new X509Certificate2(_tlsConfig.CertFile);
                    }

                    opts.AddCertificate(cert);

                    // For self-signed certificates, skip server certificate validation
                    // The connection is still encrypted with TLS
                    if (_tlsConfig.InsecureSkipVerify)
                    {
                        opts.TLSRemoteCertificationValidationCallback =
                            (sender, certificate, chain, sslPolicyErrors) =>
                            {
                                // Skip validation for self-signed certificates
                                // Connection is still encrypted with TLS
                                return true;
                            };
                    }
                }
                catch (Exception ex)
                {
                    throw new InvalidOperationException($"Failed to load TLS certificate: {ex.Message}", ex);
                }
            }

            _nc = new ConnectionFactory().CreateConnection(opts);

            string subject = $"$SRV.{_name}.>";
            _nc.SubscribeAsync(subject, HandleRPC);

            _heartbeatTimer = new Timer(_ => SendHeartbeat(), null, 0, HeartbeatIntervalMs);

            _running = true;
        }

        private void HandleRPC(object? sender, MsgHandlerEventArgs e)
        {
            var msg = e.Message;
            try
            {
                string json = System.Text.Encoding.UTF8.GetString(msg.Data);
                var request = JsonSerializer.Deserialize<RPCRequest>(json);
                if (request == null)
                {
                    SendError(msg, "", "Invalid request: null");
                    return;
                }

                _rpcLock.EnterReadLock();
                if (!_rpcHandlers.TryGetValue(request.Method, out var handler))
                {
                    _rpcLock.ExitReadLock();
                    SendError(msg, request.Id, $"Method not found: {request.Method}");
                    return;
                }
                _rpcLock.ExitReadLock();

                var result = handler(request.Args).Result;

                var response = new RPCResponse
                {
                    Id = request.Id,
                    Success = true,
                    Result = result
                };

                string respJson = JsonSerializer.Serialize(response);
                byte[] data = System.Text.Encoding.UTF8.GetBytes(respJson);
                msg.Respond(data);
            }
            catch (Exception ex)
            {
                SendError(msg, "", ex.Message);
            }
        }

        private void SendError(Msg msg, string requestId, string error)
        {
            var response = new RPCResponse
            {
                Id = requestId,
                Success = false,
                Error = error
            };
            string json = JsonSerializer.Serialize(response);
            msg.Respond(System.Text.Encoding.UTF8.GetBytes(json));
        }

        private void SendHeartbeat()
        {
            if (!_running || _nc == null) return;

            var heartbeat = new
            {
                service = _name,
                version = "1.0.0",
                timestamp = ((DateTimeOffset)DateTime.UtcNow).ToUnixTimeSeconds()
            };

            string json = JsonSerializer.Serialize(heartbeat);
            _nc.Publish($"$LL.heartbeat.{_name}", System.Text.Encoding.UTF8.GetBytes(json));
        }

        public void Stop()
        {
            if (!_running) return;
            _running = false;
            _heartbeatTimer?.Dispose();
            _nc?.Close();
            _nc = null;
        }

        // BuildCurrentMetadata - builds metadata from registered methods
        public ServiceMetadata BuildCurrentMetadata(
            string version,
            string description,
            string author,
            List<string> tags)
        {
            _metaLock.EnterReadLock();
            try
            {
                var methods = new List<MethodMetadata>();
                foreach (var meta in _methodMetadata.Values)
                {
                    methods.Add(meta);
                }

                return new ServiceMetadata
                {
                    Name = _name,
                    Version = version,
                    Description = description,
                    Author = author,
                    Tags = tags,
                    Methods = methods,
                    RegisteredAt = DateTime.UtcNow,
                    LastSeen = DateTime.UtcNow
                };
            }
            finally
            {
                _metaLock.ExitReadLock();
            }
        }

        // RegisterMetadata - publish metadata to $LL.register.{service}
        public void RegisterMetadata(ServiceMetadata metadata)
        {
            if (_nc == null)
                throw new InvalidOperationException("Service not started. Call Start() first.");

            // Get instance info
            var hostIP = GetLocalIPAddress();
            var hostMAC = GetMacAddress();
            var workingDir = Environment.CurrentDirectory;

            Console.WriteLine($"[DEBUG] Instance info - IP: {hostIP}, MAC: {hostMAC}, Dir: {workingDir}");

            var instanceInfo = new InstanceInfo
            {
                Language = "csharp",
                HostIP = hostIP,
                HostMAC = hostMAC,
                WorkingDir = workingDir
            };

            var msg = new
            {
                service = _name,
                version = metadata.Version,
                metadata = metadata,
                timestamp = ((DateTimeOffset)DateTime.UtcNow).ToUnixTimeSeconds(),
                instance_info = instanceInfo
            };

            string json = JsonSerializer.Serialize(msg);
            byte[] data = System.Text.Encoding.UTF8.GetBytes(json);
            _nc.Publish($"$LL.register.{_name}", data);
        }

        public void Dispose()
        {
            Stop();
            _rpcLock?.Dispose();
            _metaLock?.Dispose();
        }

        // GetLocalIPAddress gets the local IP address
        private string GetLocalIPAddress()
        {
            try
            {
                // Try to get IPv4 address from all network interfaces
                foreach (var nic in NetworkInterface.GetAllNetworkInterfaces())
                {
                    if (nic.OperationalStatus == OperationalStatus.Up &&
                        nic.NetworkInterfaceType != NetworkInterfaceType.Loopback)
                    {
                        var ipProps = nic.GetIPProperties();
                        foreach (var addr in ipProps.UnicastAddresses)
                        {
                            if (addr.Address.AddressFamily == AddressFamily.InterNetwork)
                            {
                                return addr.Address.ToString();
                            }
                        }
                    }
                }
            }
            catch
            {
                // Ignore errors
            }
            return "127.0.0.1";
        }

        // GetMacAddress gets the MAC address
        private string GetMacAddress()
        {
            try
            {
                foreach (var nic in NetworkInterface.GetAllNetworkInterfaces())
                {
                    if (nic.OperationalStatus == OperationalStatus.Up &&
                        nic.NetworkInterfaceType != NetworkInterfaceType.Loopback)
                    {
                        var macBytes = nic.GetPhysicalAddress().GetAddressBytes();
                        if (macBytes.Length > 0)
                        {
                            // Format as XX:XX:XX:XX:XX:XX
                            var macAddr = string.Join(":", macBytes.Select(b => b.ToString("X2")));
                            return macAddr;
                        }
                    }
                }
            }
            catch
            {
                // Ignore errors
            }
            return "00:00:00:00:00:00";
        }
    }
}
